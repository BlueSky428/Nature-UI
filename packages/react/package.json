{
  "name": "@nature-ui/react",
  "version": "0.1.0",
  "description": "Responsive and accessible React UI components built with react, tailwindcss and emotion",
  "keywords": [
    "react",
    "ui",
    "design-system",
    "react-components",
    "uikit",
    "accessible",
    "components",
    "emotion",
    "library",
    "design-system"
  ],
  "author": "Divine Hycenth <divinehycenth@outlook.com>",
  "homepage": "https://github.com/DNature/nature-ui",
  "license": "MIT",
  "main": "dist/index.js",
  "directories": {
    "lib": "dist",
    "test": "__tests__"
  },
  "files": [
    "dist",
    "src"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/DNature/nature-ui.git"
  },
  "scripts": {
    "build": "rimraf ./dist && tsc",
    "test": "jest --env=jsdom"
  },
  "bugs": {
    "url": "https://github.com/DNature/nature-ui/issues"
  },
  "dependencies": {
    "@nature-ui/alert": "^0.1.0",
    "@nature-ui/alert-dialog": "^0.1.0",
    "@nature-ui/avatar": "^0.1.0",
    "@nature-ui/breadcrumb": "^0.1.0",
    "@nature-ui/button": "^0.1.0",
    "@nature-ui/checkbox": "^0.1.0",
    "@nature-ui/progress": "^0.1.0",
    "@nature-ui/clickable": "^0.1.0",
    "@nature-ui/close-button": "^0.1.0",
    "@nature-ui/collapse": "^0.1.0",
    "@nature-ui/container": "^0.1.0",
    "@nature-ui/drawer": "^0.1.0",
    "@nature-ui/form-control": "^0.1.0",
    "@nature-ui/hooks": "^0.1.0",
    "@nature-ui/icon": "^0.1.0",
    "@nature-ui/image": "^0.1.0",
    "@nature-ui/input": "^0.1.0",
    "@nature-ui/layout": "^0.1.0",
    "@nature-ui/modal": "^0.1.0",
    "@nature-ui/popover": "^0.1.0",
    "@nature-ui/popper": "^0.1.0",
    "@nature-ui/portal": "^0.1.0",
    "@nature-ui/radio": "^0.1.0",
    "@nature-ui/spinner": "^0.1.0",
    "@nature-ui/system": "^0.1.0",
    "@nature-ui/toast": "^0.1.0",
    "@nature-ui/transition": "^0.1.0",
    "@nature-ui/utils": "^0.1.0",
    "@nature-ui/visually-hidden": "^0.1.0"
  },
  "devDependencies": {
    "@nature-ui/system": "^0.1.0",
    "rimraf": "^3.0.2"
  },
  "peerDependencies": {
    "@nature-ui/system": "^0.1.0",
    "react": "16.x",
    "rimraf": "^3.0.2"
  }
}
