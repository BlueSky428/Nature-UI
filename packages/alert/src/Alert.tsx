import * as React from 'react';
import { clsx as clx, nature, PropsOf } from '@nature-ui/system';
import {
  FiInfo,
  FiAlertCircle,
  FiAlertTriangle,
  FiCheckCircle,
} from 'react-icons/fi';
import { createContext, __DEV__ } from '@nature-ui/utils';

const SUBTLE_TEXT = 'text-gray-800';

export const ALERT_STATUSES = {
  error: {
    bg: 'bg-red-200',
    iconColor: 'text-red-600 mr-3',
    icon: FiAlertCircle,
    variant: {
      solid: 'bg-red-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  info: {
    bg: 'bg-blue-200',
    iconColor: 'text-blue-600 mr-3',
    icon: FiInfo,
    variant: {
      solid: 'bg-blue-600',
      subtle: SUBTLE_TEXT,
    },
  },
  success: {
    bg: 'bg-green-200',
    iconColor: 'text-green-600 mr-3',
    icon: FiCheckCircle,
    variant: {
      solid: 'bg-green-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
  warning: {
    bg: 'bg-orange-200',
    iconColor: 'text-orange-600 mr-3',
    icon: FiAlertTriangle,
    variant: {
      solid: 'bg-orange-600 text-white',
      subtle: SUBTLE_TEXT,
    },
  },
};

type AlertContext = Required<Pick<AlertOptions, 'status' | 'variant'>>;

const [AlertContextProvider, useAlertContext] = createContext<AlertContext>({
  name: 'AlertContext',
});

interface AlertOptions {
  /**
   * The status of the alert
   */
  status?: keyof typeof ALERT_STATUSES;
  /**
   * The variant of the alert style to use
   */
  variant?: 'solid' | 'subtle';
  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component?: React.ElementType;
  id?: string;
  alertTitle?: React.ReactNode;
}

const DivTag = nature('div');

export type AlertProps = AlertOptions & PropsOf<typeof DivTag>;

const BASE_STYLE =
  'px-4 py-3 flex items-center w-full relative overflow-hidden';

export const AlertWrapper = (props: AlertProps) => {
  const {
    className = '',
    status = 'success',
    variant = 'subtle',
    component: Component = DivTag,
    role = 'alert',
    ...rest
  } = props;

  // const Component = 'div';

  const { variant: Variant, bg } = ALERT_STATUSES[status];
  const VARIANT: string = Variant[variant];

  const componentClass = clx(className, BASE_STYLE, {
    [bg]: status,
    [VARIANT]: variant,
  });

  const context = {
    status,
    variant,
  };

  return (
    <AlertContextProvider value={context}>
      <Component className={componentClass} {...rest} role={role} />
    </AlertContextProvider>
  );
};

if (__DEV__) {
  AlertWrapper.displayName = 'Alert';
}

export type AlertTitleProps = PropsOf<typeof DivTag>;

export const AlertTitle = (props: AlertTitleProps) => {
  const { className = '', ...rest } = props;
  // const Component = 'div';

  return <DivTag className={clx(className, 'font-bold')} {...rest} />;
};

if (__DEV__) {
  AlertTitle.displayName = 'AlertTitle';
}

export type AlertDescriptionProps = PropsOf<typeof DivTag>;

export const AlertDescription = (props: AlertDescriptionProps) => {
  const { className = '', ...rest } = props;
  // const Component = 'div';

  return <DivTag className={className} {...rest} />;
};

if (__DEV__) {
  AlertDescription.displayName = 'AlertDescription';
}

const SpanTag = nature('span');

export type AlertIconProps = PropsOf<typeof SpanTag>;

export const AlertIcon = (props: AlertIconProps) => {
  const { className = '' } = props;

  const { variant = 'subtle', status = 'success' } = useAlertContext();

  // const Component = 'div';

  const { iconColor, icon: IconComponent, variant: Variant } = ALERT_STATUSES[
    status
  ];
  const VARIANT: string = Variant[variant];

  // const Icon = ALERT_STATUSES[status];

  const iconClasses = clx(className, {
    [VARIANT]: variant,
    'mr-3': variant,
    [iconColor]: variant !== 'solid',
  });

  return <IconComponent className={iconClasses} size={20} />;
};

if (__DEV__) {
  AlertIcon.displayName = 'AlertIcon';
}

export const Alert = (props: AlertProps) => {
  const {
    variant = 'subtle',
    status = 'error',
    alertTitle,
    children,
    ...rest
  } = props;

  return (
    <AlertWrapper {...rest} variant={variant} status={status}>
      <AlertIcon />
      <AlertTitle className='mr-3'>{alertTitle}</AlertTitle>
      <AlertDescription>{children}</AlertDescription>
    </AlertWrapper>
  );
};

if (__DEV__) {
  Alert.displayName = 'Alert';
}
