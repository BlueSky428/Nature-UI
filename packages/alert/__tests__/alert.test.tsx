import * as React from 'react';
import { render, screen } from '@testing-library/react';

import Alert from '../src';

describe('Alert test: ', () => {
  it('should be in the dom', () => {
    render(
      <Alert component='div' status='success'>
        Hi world
      </Alert>
    );

    expect(screen.getByRole('div')).toHaveTextContent('Hi world');
  });
});
