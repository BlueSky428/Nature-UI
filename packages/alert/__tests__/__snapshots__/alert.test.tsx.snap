// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@nature-ui/alert  should also render correctly 1`] = `
<DocumentFragment>
  <div
    class="px-4 py-3 flex items-center w-full relative overflow-hidden bg-red-200 text-gray-800"
    role="alert"
  >
    <svg
      class="text-red-600 mr-3 css-rre05l-Icon"
      fill="none"
      focusable="false"
      height="1em"
      role="presentation"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <circle
        cx="12"
        cy="12"
        r="10"
      />
      <line
        x1="12"
        x2="12"
        y1="8"
        y2="12"
      />
      <line
        x1="12"
        x2="12.01"
        y1="16"
        y2="16"
      />
    </svg>
    <p
      class="font-bold mr-3"
    >
      Alert title
    </p>
    <p>
      Alert description
    </p>
  </div>
</DocumentFragment>
`;

exports[`@nature-ui/alert  should render correctly 1`] = `
<DocumentFragment>
  <div
    class="px-4 py-3 flex items-center w-full relative overflow-hidden bg-green-200 text-gray-800"
    role="alert"
  >
    <svg
      class="text-green-600 mr-3 css-rre05l-Icon"
      fill="none"
      focusable="false"
      height="1em"
      role="presentation"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
      />
      <polyline
        points="22 4 12 14.01 9 11.01"
      />
    </svg>
    <p
      class="font-bold"
    >
      Alert title
    </p>
    <p>
      Alert description
    </p>
  </div>
</DocumentFragment>
`;
